### DISCUSSION TITLE
Recovering a Build Folder

### DISCUSSION THREAD

**Discussion by 0xWOLAND** [dennisvang/tufup#155] [Q&A] [answered]
If I deleted my pyinstaller build folder that contains all of the tufup stuff (e.g. `build/`, `dist/`, `keystore/`, ...), what is the correct way of recovering my builds. I saved my metadata files in `repository/metadata`, `keystore/*` contents, and the `.tufup-repo-config`. I noticed that the build doesn't work unless I have all of the previous `target` builds inside of the `repository/target` folder. TLDR: is there a way to recover a tufup environment if you delete all of your zipped builds. For context, I'm testing this on MacOS.

**Comment by dennisvang** [comment, marked as answer]
@0xWOLAND Good question!

As long as you've secured your `keystore` contents and your `root.json` metadata (including `*.root.json` versions), it is possible to rebuild the tufup repo.

The root metadata are important for continuity, because any apps that are already in the field will look for those root metadata, as [discussed in the TUF spec][1].

The rest of the repository can be reconstructed from there.

The important thing is that the `targets.json` metadata should match the actual content of the `repository/targets` dir.

If you still have the original target files (archives and, optionally, patches), then putting everything back in the correct folders should work.

On the other hand, if you need to rebuild your targets from scratch, it is likely that the hashes will no longer match those from your targets.json metadata file. In that case it is best to delete the old `targets.json`, `snapshot.json`, and `timestamp.json` (but NOT the `root.json` and `*.root.json` files), and start adding targets from a clean slate.

[1]: https://theupdateframework.github.io/specification/latest/#detailed-client-workflow
